### ISSUE TITLE
Transition error [page][enter] in any version after 2.3.16

### ISSUE THREAD

**Issue by carlblock** [barbajs/barba#444]
## The problem

After upgrading @barba/core to 2.6.0 our enter transition fails after navigating the web about 2-3 times.

## Description

It's reproducible on the website by going between two different pages about 3 times in row back and forth. I tried downgrading and found that 2.3.16 is the latest version where this error doesn't happen.

I also tried switching between different options such as ` cacheIgnore: true/false`. But error still persists. I'm unable to figure out why this error started happening since I have not changed any of the transitions or helper functions, I only did a update of @barba/core.

### Details

The deepest error thrown in the stacktrace seems to be this block:

```js
if (r) {
      var i = this.dom.getHref(r);
      this.prevent.checkUrl(i) || this.cache.has(i) || this.cache.set(i, this.request(i, this.timeout, this.onRequestError.bind(this, r)).catch(function(t) {
        n.logger.error(t)
      }), "enter")
    }
```

The error stops if I comment the `window.app.fetchModulesFromHtml(document)` line. However, this function didn't change since upgrading @barba/core. 

And if I go a bit deeper in my app.js code I also make @barba/core not throw an error if comment this line `window.app.runningModules.push({ moduleName: moduleName, moduleInstance: newInstance })`. But how this could affect the @barba/core is really confusing.

### Expected behaviour

No transition errors.

### Actual behaviour

@barba/core doesn't trigger the enter event, therefore causing the enter transition to fail.

### Code To Reproduce Issue

```js
const vm = this
// eslint-disable-next-line
const plugins = [ ScrollToPlugin ] // this is because gsap cant handle webpack tree-shaking
const isMobile = common.isMobile()
const htmlHeader = document.querySelector('.html-header')
barba.init({
  debug: true,
  cacheIgnore: false,
  timeout: 10000,
  transitions: [
    {
      name: 'default',
      leave ({ current, next, trigger }) {
        current.container.classList.remove('barba-container--fadein')
        current.container.classList.toggle('barba-container--fadeout')
        setTimeout(() => {
          this.async()
        }, 400)
      },
      enter ({ current, next, trigger }) {
        document.querySelector('body').classList.remove('noScroll')
        next.container.classList.remove('barba-container--fadeout')
        next.container.classList.toggle('barba-container--fadein')
      }
    },
    {
      name: 'nextwork',
      from: {
        custom: ({ trigger }) => {
          if (trigger.classList && trigger.classList.contains('work-footer__next') && isMobile === false) {
            return true
          }
          if (trigger.classList && trigger.classList.contains('work-footer__title') && isMobile === false) {
            return true
          }
        }
      },
      leave ({ current, next, trigger }) {
        const done = this.async()
        vm.nextWorkAnimation().then(() => {
          done()
        })
      },
      enter ({ current, next, trigger }) {
        // window.scrollTo(0, 0)
      }
    },
    {
      name: 'grid',
      from: {
        custom: ({ trigger }) => {
          if (window.innerWidth < 960) {
            return false
          }
          if (trigger.classList && trigger.classList.contains('SlideImgWrap-link')) {
            return true
          }
          if (trigger.classList && trigger.classList.contains('SlideImgWrap-image')) {
            return true
          }
          if (trigger.classList && trigger.classList.contains('GridMenuItem-title')) {
            return true
          }
        }
      },
      leave ({ current, next, trigger }) {
        const done = this.async()
        vm.gridAnimation().then(() => {
          done()
        })
      },
      enter ({ current, next, trigger }) {
        document.querySelector('body').classList.remove('noScroll')
      }
    }
  ],
  prevent: ({ el }) => {
    if (/.pdf/.test(el.href.toLowerCase())) {
      return true
    }
    if (el.classList && el.classList.contains('prevent')) {
      return true
    }
    if (el.dataset && el.dataset.lightbox) {
      return true
    }
  },
  requestError: (trigger, action, url, response) => {
    console.log('REQUEST ERROR')
    console.log(url)
    console.log(response)
  }
})
// Global leave hook
barba.hooks.leave(({ current, next, trigger }) => {
  vm.hideMainMenu(trigger, htmlHeader)
  vm.hideBackButton(htmlHeader)
})
// Global enter hook
barba.hooks.enter(({ current, next, trigger }) => {
  // window.scrollTo(0, 0)
  window.requestAnimationFrame(() => {
    vm.setBodyClasses(next.container)
    vm.mobileMenu()
    if (htmlHeader) htmlHeader.classList.remove('no-update')
    vm.fixWpAdminBar()
    vm.updateCurrentHighlightedMenu(next.html)
    window.app.fetchModulesFromHtml(next.html)
  })
})
```
_app.js_
```js
/* eslint-disable no-unused-vars */
import upperFirst from 'lodash-es/upperFirst'
import camelCase from 'lodash-es/camelCase'

class App {
  constructor () {
    this.modules = []
    this.runningModules = []
  }

  /**
   * Loads modules without initialising them.
   * @param {Object|Array} moduleClass
   * @param moduleName
   * @param {Object} options
   */
  addModule (moduleClass, moduleName = '', options = {}) {
    if (typeof moduleClass === 'object') {
      window.app.modules.push({ moduleClass: moduleClass, moduleName: moduleName, moduleSettings: options })
    }
  }

  /**
   * Calls the destroy function on the module class and removes it from this.runningModules
   * @param moduleName
   * @returns {Promise<any>}
   */
  destroyModuleInstance (moduleName) {
    return new Promise((resolve, reject) => {
      try {
        const moduleIndex = window.app.runningModules.findIndex(mod => mod.moduleName === moduleName)
        if (moduleIndex !== -1) {
          window.app.runningModules[moduleIndex]['moduleInstance'].destroy()
          // Remove it from running modules
          window.app.runningModules = window.app.runningModules.filter(mod => mod.moduleName !== moduleName)
          resolve(moduleName)
        }
        reject(new Error(`Could not destroy module ${moduleName}`))
      } catch (e) {
        reject(e)
      }
    })
  }

  /**
   * Creates a new instance of the loaded module and inserts it in this.runningModules
   * @param moduleName
   * @returns {Promise<any>}
   */
  initModuleInstance (moduleName) {
    console.log('moduleName:', moduleName)
    return new Promise((resolve, reject) => {
      if (this.persistentModule(moduleName) && window.app.isRunning(moduleName)) {
        reject(new Error('Not re-creating persistent module'))
      }
      try {
        // Create a new class instance from the modules container and call init()
        const moduleIndex = window.app.modules.findIndex(module => module.moduleName === moduleName)
        if (moduleIndex !== -1) {
          const newInstance = Object.create(Object.getPrototypeOf(window.app.modules[moduleIndex]['moduleClass']))
          newInstance.constructor.apply(newInstance)
          newInstance.init()
          window.app.runningModules.push({ moduleName: moduleName, moduleInstance: newInstance })
          resolve(moduleName)
        }
        reject(new Error(`Could not find module: ${moduleName}`))
      } catch (e) {
        console.log('Could not init:', moduleName)
        reject(e)
      }
    })
  }

  /**
   * Returns module settings for persistence
   * @param moduleName
   * @returns {boolean}
   */
  persistentModule (moduleName) {
    const moduleIndex = window.app.modules.findIndex(module => module.moduleName === moduleName)
    if (moduleIndex === -1) return false
    const moduleSettings = window.app.modules[moduleIndex].moduleSettings
    return 'persistent' in moduleSettings && moduleSettings.persistent === true
  }

  /**
   * Whether module is running or not
   * @param moduleName
   * @returns {boolean}
   */
  isRunning (moduleName) {
    return !!window.app.runningModules.find(module => module.moduleName === moduleName)
  }

  /**
   * Calls destroy on all running modules and removes them from this.runningModules
   * @returns {Promise<any[] | never>}
   */
  destroyModules () {
    const promises = []
    for (const module of window.app.runningModules) {
      // Don't destroy module if it is persistent type
      if (window.app.persistentModule(module.moduleName) === false) {
        promises.push(window.app.destroyModuleInstance(module.moduleName))
      }
    }
    return Promise.all(promises).then((res) => {
      // console.log('Destroyed:', res.join(', '))
    }).catch(e => {
      console.warn(e)
    })
  }

  /**
   * Takes an array of moduleNames and initialises them
   * @param modules
   * @returns {Promise<any[] | never>}
   */
  initModules (modules) {
    console.log(modules)
    const promises = []
    for (const moduleName of modules) {
      // Don't init module if its persistent and already running
      if (window.app.persistentModule(moduleName) && window.app.isRunning(moduleName) === false) {
        promises.push(window.app.initModuleInstance(moduleName))
      }
      // Non-persistent modules can be initialised
      if (window.app.persistentModule(moduleName) === false) {
        promises.push(window.app.initModuleInstance(moduleName))
      }
    }
    return Promise.all(promises).then((res) => {
      console.log('Initialised:', res.join(', '))
    }).catch(e => {
      console.warn(e)
    })
  }

  /**
   * Fetches the modules from the dom
   * @param {String|Node|Document} htmlString or document
   */
  fetchModulesFromHtml (htmlString) {
    if (typeof htmlString === 'object') {
      htmlString = new window.XMLSerializer().serializeToString(htmlString)
    }
    const html = new window.DOMParser().parseFromString(htmlString, 'text/html')
    if ('body' in html === false || html.body.childNodes.length < 1) return
    const modulesList = html.querySelector('.data')
    if (modulesList !== null) {
      const modules = modulesList.dataset.modules.split(',')
      const moduleNames = modules.map(module => upperFirst(camelCase(module)))
      window.app.destroyModules().then(() => {
        window.app.initModules(moduleNames).then(() => {
          // All modules initialised
        })
      })
    }
  }
}

export default App
```

## Environment

- Barba version that exhibits the issue:
- Build used:
  - 2.6.0 @barba/core
- Browser OS/version used:
  - Chrome
  - Safari
  - Firefox

## Probably related issues
#391

**Comment by geanfrancovolpe** [comment]
Have you solved that problem bro?

**Comment by carlblock** [comment]
Hey @geanfrancovolpe, I haven't solved it yet. I've decided to downgrade to latest version before 2.3.16 instead. Hopefully I will have some spare time next week to look deeper into this issue.

**Comment by xavierfoucrier** [comment]
Hi @carlblock,

Waiting for your feedback regarding this issue.
Big thanks for your patience :+1:

**Comment by thierrymichel** [comment]
Hi @carlblock,

Sorry for the delay…

> I tried downgrading and found that 2.3.16 is the latest version where this error doesn't happen.

So can you confirm that it should be "somewhere" here: https://github.com/barbajs/barba/compare/@barba/core@2.3.16...@barba/core@2.4.0 :D

**Comment by carlblock** [comment]
Hey @thierrymichel 

Sorry for the delays, I've had time to finally look at this issue again, I think I nailed the problem being caused [because of this.](https://github.com/barbajs/barba/compare/@barba/core@2.3.16...@barba/core@2.6.0#diff-303d7832d81c8ba43fd8a5b228edd981R86-R106)

But however the good news is that I believe [this commit](https://github.com/barbajs/barba/commit/7f349a7b718df3666d9a626a6902b5d934b29726#diff-303d7832d81c8ba43fd8a5b228edd981) from you was the fix to this issue.

I just tried replicating the issue with 2.8.0 and everything seems to work fine so far!

Sorry again for the delay, I'm having some busy days!

Feel fine to close this issue and thanks for the hard work on barbajs!
